File: showcase/integrations/_shared/dotnet/tests/CvdiagSchema.test.cs
// CvdiagSchema.test.cs — xUnit red-green proof for the .NET CVDIAG codegen
// binding (plan unit L0-D, spec §5/§6). Mirrors the TS `schema.test.ts` /
// `emit.ts` contract:
//
//   1. schema round-trip — a fully-populated CvdiagEnvelope serializes to JSON
//      and deserializes back identically (System.Text.Json source-gen).
//   2. forbidden-header rejection — CvdiagEmitter.FilterEdgeHeaders() drops a
//      DENY-list header (cf-ipcountry) and the cf-ip* family even if it is
//      injected alongside an allow-list header; every result carries all 9
//      allow-list keys (absent → null).
//   3. _metadata_dropped stamp — an envelope carrying an unknown metadata key
//      is closed-world filtered and the `_metadata_dropped` flag is set.
//   4. production-env DEBUG refusal — CvdiagEmitter fail-closes (throws) when
//      DEBUG is requested in a production environment, when no env resolves,
//      and when the allow-list is absent.
//
// RED: with no CvdiagSchema.cs / CvdiagEmitter.cs these types do not exist and
// the suite does not compile. GREEN once both are implemented.
//
// Run (from repo root, requires the .NET 9 SDK):
//   dotnet test showcase/integrations/_shared/dotnet/tests/CvdiagSchema.Tests.csproj

using System.Text.Json;
using Copilotkit.Showcase.Cvdiag;
using Xunit;

namespace Copilotkit.Showcase.Cvdiag.Tests;

public class CvdiagSchemaTests
{
    private static EdgeHeaders EmptyEdgeHeaders() => new();

    private static CvdiagEnvelope SampleEnvelope() => new()
    {
        SchemaVersion = CvdiagSchema.SchemaVersion,
        TestId = "017f22e2-79b0-7cc3-98c4-dc0c0c07398f",
        TraceId = "017f22e2-79b0-7cc3-98c4-dc0c0c07398f",
        SpanId = "00f067aa0ba902b7",
        ParentSpanId = null,
        Layer = CvdiagLayer.Backend,
        Boundary = CvdiagBoundary.BackendResponseComplete,
        Slug = "gen-ui-chat",
        Demo = "default",
        Ts = "2026-06-18T12:00:00.000Z",
        MonoNs = 123456789,
        DurationMs = 42,
        Outcome = CvdiagOutcome.Ok,
        EdgeHeaders = EmptyEdgeHeaders(),
        Metadata = new Dictionary<string, object?> { ["http_status"] = 200 },
    };

    // (1) Round-trip: serialize → deserialize → equal field-by-field.
    [Fact]
    public void Envelope_RoundTrips_ThroughSystemTextJson()
    {
        var original = SampleEnvelope();
        var json = JsonSerializer.Serialize(original, CvdiagJsonContext.Default.CvdiagEnvelope);
        var back = JsonSerializer.Deserialize(json, CvdiagJsonContext.Default.CvdiagEnvelope)!;

        Assert.Equal(original.SchemaVersion, back.SchemaVersion);
        Assert.Equal(original.TestId, back.TestId);
        Assert.Equal(original.SpanId, back.SpanId);
        Assert.Equal(original.Layer, back.Layer);
        Assert.Equal(original.Boundary, back.Boundary);
        Assert.Equal(original.Slug, back.Slug);
        Assert.Equal(original.Outcome, back.Outcome);
        Assert.Equal(original.MonoNs, back.MonoNs);
        Assert.Equal(original.DurationMs, back.DurationMs);
    }

    // (1b) Wire format: enum + key names match the schema.json contract exactly
    // (snake_case keys, dotted/lowercase enum string values).
    [Fact]
    public void Envelope_SerializesWithSchemaWireNames()
    {
        var json = JsonSerializer.Serialize(SampleEnvelope(), CvdiagJsonContext.Default.CvdiagEnvelope);
        Assert.Contains("\"schema_version\":1", json);
        Assert.Contains("\"test_id\":", json);
        Assert.Contains("\"parent_span_id\":null", json);
        Assert.Contains("\"edge_headers\":", json);
        Assert.Contains("\"layer\":\"backend\"", json);
        Assert.Contains("\"boundary\":\"backend.response.complete\"", json);
        Assert.Contains("\"outcome\":\"ok\"", json);
        // cf-ray etc. serialize with their hyphenated wire names.
        Assert.Contains("\"cf-ray\":null", json);
        Assert.Contains("\"x-railway-request-id\":null", json);
    }

    // (2) Forbidden-header rejection: cf-ipcountry (DENY) is dropped even when
    // injected next to an allow-list header; the cf-ip* family is never
    // captured; all 9 allow-list keys are present (absent → null).
    [Fact]
    public void FilterEdgeHeaders_RejectsDenyListAndCfIpFamily()
    {
        var raw = new Dictionary<string, string?>
        {
            ["cf-ray"] = "8abc-DFW",                 // allow → kept
            ["cf-ipcountry"] = "US",                 // deny → dropped
            ["cf-connecting-ip"] = "1.2.3.4",        // deny → dropped
            ["true-client-ip"] = "1.2.3.4",          // deny → dropped
            ["x-forwarded-for"] = "1.2.3.4",         // deny → dropped
            ["CF-IPCity"] = "Dallas",                // deny (case-insensitive)
            ["server"] = "railway-edge",             // allow → kept
        };

        var filtered = CvdiagEmitter.FilterEdgeHeaders(raw);

        Assert.Equal("8abc-DFW", filtered.CfRay);
        Assert.Equal("railway-edge", filtered.Server);
        // DENY-list values never surface on any of the 9 fields.
        Assert.DoesNotContain("US", filtered.AllValues());
        Assert.DoesNotContain("1.2.3.4", filtered.AllValues());
        Assert.DoesNotContain("Dallas", filtered.AllValues());
        // All 9 allow-list keys present; unset → null.
        Assert.Null(filtered.CfMitigated);
        Assert.Null(filtered.XRailwayRequestId);
        Assert.Null(filtered.RetryAfter);
        Assert.Equal(9, filtered.KeyCount());
    }

    // (2b) Deny wins even if a deny key somehow appears in the allow path.
    [Fact]
    public void FilterEdgeHeaders_DenyWinsOverAllow()
    {
        var raw = new Dictionary<string, string?> { ["forwarded"] = "for=1.2.3.4" };
        var filtered = CvdiagEmitter.FilterEdgeHeaders(raw);
        Assert.DoesNotContain("for=1.2.3.4", filtered.AllValues());
    }

    // (3) _metadata_dropped stamp: an unknown metadata key for the boundary is
    // dropped (closed-world) and the flag is stamped.
    [Fact]
    public void Emit_StampsMetadataDropped_OnUnknownKey()
    {
        var emitter = CvdiagEmitterTestFactory.Verbose();
        var env = emitter.Emit(new CvdiagEmitArgs
        {
            Layer = CvdiagLayer.Backend,
            Boundary = CvdiagBoundary.BackendResponseComplete,
            Slug = "gen-ui-chat",
            Demo = "default",
            Outcome = CvdiagOutcome.Ok,
            Metadata = new Dictionary<string, object?>
            {
                ["http_status"] = 200,        // allowed for this boundary
                ["totally_unknown_key"] = 1,  // not in the closed key set → dropped
            },
        });

        Assert.NotNull(env);
        Assert.True(env!.MetadataDropped);
        Assert.True(env.Metadata.ContainsKey("http_status"));
        Assert.False(env.Metadata.ContainsKey("totally_unknown_key"));
    }

    // (3b) Clean metadata leaves the flag unset.
    [Fact]
    public void Emit_DoesNotStampMetadataDropped_OnCleanMetadata()
    {
        var emitter = CvdiagEmitterTestFactory.Verbose();
        var env = emitter.Emit(new CvdiagEmitArgs
        {
            Layer = CvdiagLayer.Backend,
            Boundary = CvdiagBoundary.BackendAgentEnter,
            Slug = "gen-ui-chat",
            Demo = "default",
            Outcome = CvdiagOutcome.Ok,
            Metadata = new Dictionary<string, object?>
            {
                ["agent_name"] = "proverbs",
                ["model_id"] = "gpt-4o-mini",
            },
        });

        Assert.NotNull(env);
        Assert.False(env!.MetadataDropped);
    }

    // (4) Production-env DEBUG refusal: SHOWCASE_ENV=production + DEBUG throws.
    [Fact]
    public void DebugInProduction_FailsClosed()
    {
        var env = new Dictionary<string, string?>
        {
            ["SHOWCASE_ENV"] = "production",
            ["CVDIAG_DEBUG_ALLOW_LIST"] = "gen-ui-chat",
        };
        Assert.Throws<InvalidOperationException>(
            () => new CvdiagEmitter(new CvdiagEmitterOptions { Debug = true, Env = env }));
    }

    // (4b) Unresolved env (no SHOWCASE_ENV/RAILWAY/ASPNETCORE) treated as prod.
    [Fact]
    public void DebugWithUnresolvedEnv_FailsClosed()
    {
        var env = new Dictionary<string, string?> { ["CVDIAG_DEBUG_ALLOW_LIST"] = "gen-ui-chat" };
        Assert.Throws<InvalidOperationException>(
            () => new CvdiagEmitter(new CvdiagEmitterOptions { Debug = true, Env = env }));
    }

    // (4c) DEBUG in a non-prod env without an allow-list still refuses.
    [Fact]
    public void DebugWithoutAllowList_FailsClosed()
    {
        var env = new Dictionary<string, string?> { ["SHOWCASE_ENV"] = "staging" };
        Assert.Throws<InvalidOperationException>(
            () => new CvdiagEmitter(new CvdiagEmitterOptions { Debug = true, Env = env }));
    }

    // (4d) DEBUG in a non-prod env WITH an allow-list is permitted (no throw).
    [Fact]
    public void DebugInStagingWithAllowList_IsAllowed()
    {
        var env = new Dictionary<string, string?>
        {
            ["SHOWCASE_ENV"] = "staging",
            ["CVDIAG_DEBUG_ALLOW_LIST"] = "gen-ui-chat",
        };
        var emitter = new CvdiagEmitter(new CvdiagEmitterOptions { Debug = true, Env = env });
        Assert.Equal(CvdiagTier.Debug, emitter.Tier);
    }

    // (4e) Tier env precedence: ASPNETCORE_ENVIRONMENT is the last fallback
    // (the .NET analogue of NODE_ENV).
    [Fact]
    public void ResolveEnvLabel_FallsBackToAspNetCoreEnvironment()
    {
        var env = new Dictionary<string, string?> { ["ASPNETCORE_ENVIRONMENT"] = "Production" };
        Assert.Equal("production", CvdiagEmitter.ResolveEnvLabel(env));
    }

    [Fact]
    public void ResolveEnvLabel_ShowcaseEnvWins()
    {
        var env = new Dictionary<string, string?>
        {
            ["SHOWCASE_ENV"] = "staging",
            ["RAILWAY_ENVIRONMENT_NAME"] = "production",
            ["ASPNETCORE_ENVIRONMENT"] = "Production",
        };
        Assert.Equal("staging", CvdiagEmitter.ResolveEnvLabel(env));
    }

    // Codegen coverage: all 29 metadata-bearing boundaries have a closed key set.
    [Fact]
    public void AllBoundaries_HaveClosedMetadataKeySets()
    {
        Assert.Equal(33, CvdiagSchema.AllBoundaries.Count);
        Assert.Equal(29, CvdiagSchema.BoundaryMetadataKeys.Count);
        Assert.Equal(9, CvdiagSchema.EdgeHeaderKeys.Count);
    }
}

// Small factory: a VERBOSE-tier emitter with no PB writer for metadata tests.
internal static class CvdiagEmitterTestFactory
{
    public static CvdiagEmitter Verbose() => new(new CvdiagEmitterOptions
    {
        Verbose = true,
        Env = new Dictionary<string, string?> { ["SHOWCASE_ENV"] = "staging" },
    });
}

File: showcase/integrations/spring-ai/src/test/java/com/copilotkit/showcase/springai/cvdiag/CvdiagSchemaTest.java
package com.copilotkit.showcase.springai.cvdiag;

import com.copilotkit.showcase.springai.cvdiag.CvdiagSchema.CvdiagBoundary;
import com.copilotkit.showcase.springai.cvdiag.CvdiagSchema.CvdiagEnvelope;
import com.copilotkit.showcase.springai.cvdiag.CvdiagSchema.CvdiagLayer;
import com.copilotkit.showcase.springai.cvdiag.CvdiagSchema.CvdiagOutcome;
import com.copilotkit.showcase.springai.cvdiag.CvdiagSchema.EdgeHeaders;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * L0-E foundation tests for the inlined Java CVDIAG binding + emitter
 * (single-module spring-ai, no Maven reactor — decision D4).
 *
 * <p>These mirror the canonical TS gate tests in
 * {@code showcase/harness/src/cvdiag/schema.test.ts} so the Java binding
 * codegen'd from {@code showcase/harness/src/cvdiag/schema.json} pins the same
 * contract:
 * <ol>
 *   <li>schema round-trip (envelope serializes through Jackson and decodes
 *       back with all closed fields intact),</li>
 *   <li>a DENY-list edge header is rejected even if it collides with an
 *       allow-list key (exact-match deny wins),</li>
 *   <li>an unknown per-boundary metadata key is dropped and the survivor is
 *       stamped {@code _metadata_dropped: true},</li>
 *   <li>the DEBUG tier refuses to start when the resolved environment is
 *       production (fail-closed).</li>
 * </ol>
 */
class CvdiagSchemaTest {

    private final ObjectMapper mapper = new ObjectMapper();

    // (1) schema round-trip — a fully-populated envelope serializes and
    // deserializes losslessly through Jackson, with all 9 edge-header keys
    // present and the closed enums round-tripping by their wire literal.
    @Test
    void envelopeRoundTripsThroughJackson() throws Exception {
        EdgeHeaders edge = new EdgeHeaders(
                "cf-ray-1", null, "HIT", null, "req-9", null, null, "1.1 cloudflare", "railway");
        Map<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("delta_ms_from_start", 12);
        metadata.put("text_length", 4);

        CvdiagEnvelope envelope = new CvdiagEnvelope(
                CvdiagSchema.SCHEMA_VERSION,
                "017f22e2-79b0-7cc3-98c4-dc0c0c07398f",
                "017f22e2-79b0-7cc3-98c4-dc0c0c07398f",
                "0123456789abcdef",
                null,
                CvdiagLayer.PROBE,
                CvdiagBoundary.PROBE_DOM_FIRSTTOKEN,
                "langgraph-python",
                "agentic_chat",
                "2026-06-18T00:00:00.000Z",
                1L,
                null,
                CvdiagOutcome.OK,
                edge,
                metadata,
                null,
                null);

        String json = mapper.writeValueAsString(envelope);
        JsonNode node = mapper.readTree(json);

        // Closed enums serialize by their wire literal, not the Java name.
        assertThat(node.path("layer").asText()).isEqualTo("probe");
        assertThat(node.path("boundary").asText()).isEqualTo("probe.dom.firsttoken");
        assertThat(node.path("outcome").asText()).isEqualTo("ok");
        assertThat(node.path("schema_version").asInt()).isEqualTo(1);
        // All 9 edge-header keys are present (absent ones serialize as null).
        JsonNode edgeNode = node.path("edge_headers");
        assertThat(edgeNode.has("cf-ray")).isTrue();
        assertThat(edgeNode.path("cf-ray").asText()).isEqualTo("cf-ray-1");
        assertThat(edgeNode.has("cf-mitigated")).isTrue();
        assertThat(edgeNode.path("cf-mitigated").isNull()).isTrue();
        assertThat(edgeNode.path("via").asText()).isEqualTo("1.1 cloudflare");
        assertThat(edgeNode.path("server").asText()).isEqualTo("railway");
        // Round-trips back into a record with the enums intact.
        CvdiagEnvelope decoded = mapper.readValue(json, CvdiagEnvelope.class);
        assertThat(decoded.layer()).isEqualTo(CvdiagLayer.PROBE);
        assertThat(decoded.boundary()).isEqualTo(CvdiagBoundary.PROBE_DOM_FIRSTTOKEN);
        assertThat(decoded.outcome()).isEqualTo(CvdiagOutcome.OK);
        assertThat(decoded.testId()).isEqualTo("017f22e2-79b0-7cc3-98c4-dc0c0c07398f");
    }

    // The boundary enum mirrors the canonical 33 literals (29 data-plane + 4
    // accounting) from schema.json; a drift here means the codegen is stale.
    @Test
    void boundaryEnumHasThirtyThreeLiterals() {
        assertThat(CvdiagBoundary.values()).hasSize(33);
        assertThat(CvdiagBoundary.fromWire("cvdiag.metadata_dropped"))
                .isEqualTo(CvdiagBoundary.CVDIAG_METADATA_DROPPED);
        assertThat(CvdiagBoundary.fromWire("backend.llm.call.response"))
                .isEqualTo(CvdiagBoundary.BACKEND_LLM_CALL_RESPONSE);
    }

    // (2) forbidden-header rejection — a deny-list header is rejected even when
    // it collides with an allow-list key (exact-match deny wins, §5 PII).
    @Test
    void filterEdgeHeadersRejectsDenyListEvenIfCollidingWithAllowList() {
        Map<String, String> raw = new HashMap<>();
        raw.put("cf-ray", "abc-123");
        raw.put("cf-connecting-ip", "203.0.113.7"); // DENY (PII)
        raw.put("cf-ipcountry", "US"); // DENY (PII)
        raw.put("x-forwarded-for", "203.0.113.7"); // DENY (PII)
        raw.put("Via", "1.1 cloudflare"); // case-insensitive allow-list match

        EdgeHeaders filtered = CvdiagSchema.filterEdgeHeaders(raw);

        assertThat(filtered.cfRay()).isEqualTo("abc-123");
        assertThat(filtered.via()).isEqualTo("1.1 cloudflare");
        // Absent allow-list keys are present-and-null.
        assertThat(filtered.retryAfter()).isNull();
        assertThat(filtered.server()).isNull();
        // No PII value leaks onto any captured field.
        String serialized = filtered.toString();
        assertThat(serialized).doesNotContain("203.0.113.7");
        assertThat(serialized).doesNotContain("US");
    }

    // (3) unknown metadata key dropped + _metadata_dropped stamp.
    @Test
    void validateMetadataDropsUnknownKeyAndStampsDropped() {
        Map<String, Object> meta = new LinkedHashMap<>();
        meta.put("delta_ms_from_start", 10);
        meta.put("text_length", 4);
        meta.put("bogus_field", "drop-me");

        CvdiagSchema.MetadataValidationResult result =
                CvdiagSchema.validateMetadata(
                        CvdiagLayer.PROBE, CvdiagBoundary.PROBE_DOM_FIRSTTOKEN, meta);

        assertThat(result.metadata()).containsEntry("delta_ms_from_start", 10);
        assertThat(result.metadata()).doesNotContainKey("bogus_field");
        assertThat(result.metadataDropped()).isTrue();
        assertThat(result.droppedKeys()).containsExactly("bogus_field");
    }

    // The emit path stamps _metadata_dropped onto the envelope it builds when
    // the metadata validator dropped a key (end-to-end of the closed-world
    // contract, not just the validator in isolation).
    @Test
    void emitterStampsMetadataDroppedOnEnvelope() {
        Map<String, String> env = new HashMap<>();
        env.put("SHOWCASE_ENV", "staging");
        CvdiagEmitter emitter = new CvdiagEmitter(new CvdiagEmitter.Options().env(env));

        Map<String, Object> meta = new LinkedHashMap<>();
        meta.put("delta_ms_from_start", 10);
        meta.put("text_length", 4);
        meta.put("bogus_field", "drop-me");

        CvdiagEnvelope envelope = emitter.emitEvent(new CvdiagEmitter.EmitArgs()
                .layer(CvdiagLayer.PROBE)
                .boundary(CvdiagBoundary.PROBE_DOM_FIRSTTOKEN)
                .slug("langgraph-python")
                .demo("agentic_chat")
                .outcome(CvdiagOutcome.OK)
                .metadata(meta));

        assertThat(envelope).isNotNull();
        assertThat(envelope.metadataDropped()).isTrue();
        assertThat(envelope.metadata()).doesNotContainKey("bogus_field");
    }

    // (4) production-env DEBUG refusal — the emitter constructor fails closed
    // when the resolved environment label is production.
    @Test
    void debugRefusesWhenEnvironmentIsProduction() {
        Map<String, String> env = new HashMap<>();
        env.put("SHOWCASE_ENV", "production");
        assertThatThrownBy(() ->
                new CvdiagEmitter(new CvdiagEmitter.Options().debug(true).env(env)))
                .isInstanceOf(IllegalStateException.class)
                .hasMessageContaining("production");
    }

    // Fail-closed: DEBUG also refuses when NO environment label resolves at all
    // (unknown env is treated as production). Mirrors the TS gate test.
    @Test
    void debugRefusesWhenNoEnvironmentResolves() {
        assertThatThrownBy(() ->
                new CvdiagEmitter(new CvdiagEmitter.Options().debug(true).env(new HashMap<>())))
                .isInstanceOf(IllegalStateException.class);
    }

    // DEBUG is permitted on a non-production label with an allow-list present.
    @Test
    void debugAllowedOnNonProductionWithAllowList() {
        Map<String, String> env = new HashMap<>();
        env.put("SHOWCASE_ENV", "staging");
        env.put("CVDIAG_DEBUG_ALLOW_LIST", "langgraph-python");
        CvdiagEmitter emitter = new CvdiagEmitter(
                new CvdiagEmitter.Options().debug(true).env(env));
        assertThat(emitter.tier()).isEqualTo(CvdiagEmitter.Tier.DEBUG);
    }

    // Tier resolution honors the Java-specific env precedence
    // SHOWCASE_ENV -> RAILWAY_ENVIRONMENT_NAME -> SPRING_PROFILES_ACTIVE.
    @Test
    void resolveEnvLabelFollowsSpringPrecedence() {
        Map<String, String> railway = new HashMap<>();
        railway.put("RAILWAY_ENVIRONMENT_NAME", "Staging");
        assertThat(CvdiagEmitter.resolveEnvLabel(railway)).isEqualTo("staging");

        Map<String, String> springProfile = new HashMap<>();
        springProfile.put("SPRING_PROFILES_ACTIVE", "PROD");
        assertThat(CvdiagEmitter.resolveEnvLabel(springProfile)).isEqualTo("prod");

        // SHOWCASE_ENV wins over the others when all are present.
        Map<String, String> all = new HashMap<>();
        all.put("SHOWCASE_ENV", "Dev");
        all.put("RAILWAY_ENVIRONMENT_NAME", "staging");
        all.put("SPRING_PROFILES_ACTIVE", "prod");
        assertThat(CvdiagEmitter.resolveEnvLabel(all)).isEqualTo("dev");

        assertThat(CvdiagEmitter.resolveEnvLabel(new HashMap<>())).isNull();
    }
}
